File: src/analyze_includes/evaluate_includes.py
from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from src.analyze_includes.result import Result
from src.analyze_includes.system_under_inspection import (
    CcTarget,
    SystemUnderInspection,
    UsageStatus,
)

if TYPE_CHECKING:
    from src.analyze_includes.parse_source import Include


def does_include_match_available_files(
    include_statement: str, include_paths: list[str], header_files: list[str]
) -> bool:
    for header in header_files:
        for inc in include_paths:
            possible_file = inc + "/" + include_statement if inc else include_statement
            if possible_file == header:
                return True
    return False


def _include_resolves_to_any_file(included_path: Path, files: list[str]) -> bool:
    return any(included_path == Path(file).resolve() for file in files)


def _is_relative_include(
    target_under_inspection: CcTarget,
    include: Include,
    dependencies: list[CcTarget],
    include_paths: list[str],
    usage: UsageStatus,
) -> bool:
    roots_for_relative_includes = [Path(root) for root in [str(include.file.parent), *include_paths]]
    for root in roots_for_relative_includes:
        path_matching_include_statement = (root / include.include).resolve()

        # Relative include to target under inspection
        if _include_resolves_to_any_file(
            included_path=path_matching_include_statement, files=target_under_inspection.header_files
        ):
            return True

        # Relative include to dependency
        for dep in dependencies:
            if _include_resolves_to_any_file(included_path=path_matching_include_statement, files=dep.header_files):
                dep.usage.update(usage)
                return True

    return False


def _check_for_invalid_includes(
    includes: list[Include],
    dependencies: list[CcTarget],
    usage: UsageStatus,
    target_under_inspection: CcTarget,
    include_paths: list[str],
) -> list[Include]:
    invalid_includes = []

    for inc in includes:
        legal_include = False
        for dep in dependencies:
            if does_include_match_available_files(
                include_statement=inc.include,
                include_paths=include_paths,
                header_files=dep.header_files,
            ):
                legal_include = True
                dep.usage.update(usage)
                break
        if not legal_include:
            # Might be a file from the target under inspection
            legal_include = does_include_match_available_files(
                include_statement=inc.include,
                include_paths=include_paths,
                header_files=target_under_inspection.header_files,
            )
        if not legal_include:
            # Might be a relative include
            legal_include = _is_relative_include(
                target_under_inspection=target_under_inspection,
                include=inc,
                dependencies=dependencies,
                include_paths=include_paths,
                usage=usage,
            )
        if not legal_include:
            invalid_includes.append(inc)

    return invalid_includes


def _check_for_unused_dependencies(dependencies: list[CcTarget]) -> list[str]:
    return [dep.name for dep in dependencies if not dep.usage.is_used()]


def _check_for_public_deps_which_should_be_private(dependencies: SystemUnderInspection) -> list[str]:
    return [dep.name for dep in dependencies.deps if dep.usage.usage == UsageStatus.PRIVATE]


def _filter_empty_dependencies(system_under_inspection: SystemUnderInspection) -> SystemUnderInspection:
    """
    Some dependencies contain no headers and provide only libraries to link against. Since our analysis is based on
    includes we are not interested in those and throw them away to prevent them raising findings regarding unused
    dependencies.
    """
    return SystemUnderInspection(
        target_under_inspection=system_under_inspection.target_under_inspection,
        deps=[dep for dep in system_under_inspection.deps if dep.header_files],
        impl_deps=[dep for dep in system_under_inspection.impl_deps if dep.header_files],
        include_paths=system_under_inspection.include_paths,
        defines=system_under_inspection.defines,
    )


def evaluate_includes(
    public_includes: list[Include],
    private_includes: list[Include],
    system_under_inspection: SystemUnderInspection,
    ensure_private_deps: bool,
) -> Result:
    result = Result(system_under_inspection.target_under_inspection.name)
    system_under_inspection = _filter_empty_dependencies(system_under_inspection)

    result.public_includes_without_dep = _check_for_invalid_includes(
        includes=public_includes,
        dependencies=system_under_inspection.deps,
        usage=UsageStatus.PUBLIC,
        target_under_inspection=system_under_inspection.target_under_inspection,
        include_paths=system_under_inspection.include_paths,
    )
    result.private_includes_without_dep = _check_for_invalid_includes(
        includes=private_includes,
        dependencies=system_under_inspection.deps + system_under_inspection.impl_deps,
        usage=UsageStatus.PRIVATE,
        target_under_inspection=system_under_inspection.target_under_inspection,
        include_paths=system_under_inspection.include_paths,
    )

    result.unused_deps = _check_for_unused_dependencies(system_under_inspection.deps)
    result.unused_impl_deps = _check_for_unused_dependencies(system_under_inspection.impl_deps)

    if ensure_private_deps:
        result.deps_which_should_be_private = _check_for_public_deps_which_should_be_private(system_under_inspection)
        result.use_impl_deps = True

    return result

File: dwyu/aspect/private/analyze_includes/evaluate_includes.cpp
#include "dwyu/aspect/private/analyze_includes/evaluate_includes.h"

#include "dwyu/aspect/private/analyze_includes/include_statement.h"
#include "dwyu/aspect/private/analyze_includes/result.h"
#include "dwyu/aspect/private/analyze_includes/system_under_inspection.h"

#include <set>
#include <string>
#include <tuple>
#include <unordered_set>
#include <utility>
#include <vector>

namespace dwyu {
namespace {

std::vector<IncludeStatement>
findIncludesWithoutDirectDependency(const std::vector<IncludeStatement>& includes,
                                    const TargetUsage::Status usage,
                                    const std::unordered_set<std::string>& own_header_files,
                                    SystemUnderInspection::HeadersToDepsMap& direct_deps) {
    std::vector<IncludeStatement> invalid_includes{};

    for (const auto& include : includes) {
        auto dep_hdr_match = direct_deps.find(include.resolved_include);
        if (dep_hdr_match != direct_deps.end()) {
            for (auto& dep : dep_hdr_match->second) {
                dep->usage.update(usage);
            }
            continue;
        }

        const auto self_hdr_match = own_header_files.find(include.resolved_include);
        if (self_hdr_match != own_header_files.end()) {
            continue;
        }

        invalid_includes.push_back(include);
    }

    return invalid_includes;
}

} // namespace

Result evaluateIncludes(const std::vector<IncludeStatement>& public_includes,
                        const std::vector<IncludeStatement>& private_includes,
                        SystemUnderInspection& system_under_inspection,
                        const bool optimize_impl_deps) {
    Result result{system_under_inspection.target_under_inspection.name, optimize_impl_deps};

    auto public_includes_without_direct_dep = findIncludesWithoutDirectDependency(
        public_includes, TargetUsage::Status::Public, system_under_inspection.target_under_inspection.header_files,
        system_under_inspection.headers_to_public_deps_map);
    result.setPublicIncludesWithoutDirectDep(std::move(public_includes_without_direct_dep));

    auto private_includes_without_direct_dep = findIncludesWithoutDirectDependency(
        private_includes, TargetUsage::Status::Private, system_under_inspection.target_under_inspection.header_files,
        system_under_inspection.headers_to_all_deps_map);
    result.setPrivateIncludesWithoutDirectDep(std::move(private_includes_without_direct_dep));

    std::set<std::string> unused_deps{};
    for (const auto& dep_pair : system_under_inspection.headers_to_public_deps_map) {
        for (const auto& dep : dep_pair.second) {
            if (!dep->usage.is_used()) {
                std::ignore = unused_deps.insert(dep->name);
            }
        }
    }
    std::set<std::string> unused_impl_deps{};
    for (const auto& dep_pair : system_under_inspection.headers_to_all_deps_map) {
        for (const auto& dep : dep_pair.second) {
            if (!dep->usage.is_used()) {
                // Only add deps which are not yet reported as unused public deps
                if (unused_deps.find(dep->name) == unused_deps.end()) {
                    std::ignore = unused_impl_deps.insert(dep->name);
                }
            }
        }
    }

    result.setUnusedDeps(std::move(unused_deps));
    result.setUnusedImplDeps(std::move(unused_impl_deps));

    std::set<std::string> deps_which_should_be_private{};
    if (optimize_impl_deps) {
        for (const auto& dep_pair : system_under_inspection.headers_to_public_deps_map) {
            for (const auto& dep : dep_pair.second) {
                if (dep->usage.usage() == TargetUsage::Status::Private) {
                    std::ignore = deps_which_should_be_private.insert(dep->name);
                }
            }
        }
        result.setDepsWhichShouldBePrivate(std::move(deps_which_should_be_private));
    }

    return result;
}

} // namespace dwyu
